### PR TITLE
[Rest Server] Support port resources in rest server

### PR DESCRIPTION
**Pull request by abuccts**
Support port resources in rest server.

### CODE REVIEW

**Comment by qinchen123** [comment]
the README also need update.

File: job-tutorial/README.md [hunk 1]
```diff
@@ -104,6 +111,10 @@ Below please find the detailed explanation for each of the parameters in the con
 | `taskRole.cpuNumber`           | Integer, required          | CPU number for one task in the task role, no less than 1 |
 | `taskRole.memoryMB`            | Integer, required          | Memory for one task in the task role, no less than 100 |
 | `taskRole.gpuNumber`           | Integer, required          | GPU number for one task in the task role, no less than 0 |
+| `taskRole.portList`            | List, optional             | List of `portType` to use                |
+| `taskRole.portType.label`      | String in `^[A-Za-z0-9\-._~]+$` format, required | Label name for the port type |
+| `taskRole.portType.startAt`    | Integer, required          | The port no to start with in the port type, 0 for random selection |
+| `taskRole.portType.portNumber` | Integer, required          | The number of port in the port type      |
```
Comment (hwuu) [inline job-tutorial/README.md]: Number of ports of the specific type.

File: rest-server/src/config/job.js [hunk 2]
```diff
@@ -54,6 +54,31 @@ const jobConfigSchema = Joi.object().keys({
...
+          .optional()
+          .default([
+            {
+              label: 'http',
+              startAt: 0,
+              portNumber: 1,
+            },
+            {
+              label: 'ssh',
+              startAt: 0,
```
Comment (hwuu) [inline rest-server/src/config/job.js]: 22?

File: rest-server/src/config/job.js [hunk 3]
```diff
@@ -54,6 +54,31 @@ const jobConfigSchema = Joi.object().keys({
...
+              .default(0),
+            portNumber: Joi.number()
+              .integer()
+              .default(1),
+          }))
+          .optional()
+          .default([
+            {
+              label: 'http',
+              startAt: 0,
```
Comment (hwuu) [inline rest-server/src/config/job.js]: 80?

File: job-tutorial/README.md [hunk 4]
```diff
@@ -79,6 +79,13 @@ A json file describe detailed configuration required for a job submission. The d
       "cpuNumber":  Integer,
       "memoryMB":   Integer,
       "gpuNumber":  Integer,
+      "portList": [
+        {
+          "label":      String,
+          "startAt":    Integer,
```
Comment (hwuu) [inline job-tutorial/README.md]: begin / count?

File: rest-server/src/config/job.js (hunk 3 above)
Comment (qinchen123) [reply rest-server/src/config/job.js]: startAt 0 means any ports, the Launcher will select one.

---
In reply to: [176960001](https://github.com/Microsoft/pai/pull/409#discussion_r176960001) [](ancestors = 176960001)

File: rest-server/src/config/job.js (hunk 2 above)
Comment (qinchen123) [reply rest-server/src/config/job.js]: startAt 0 means any ports, the Launcher will select one

---
In reply to: [176959993](https://github.com/Microsoft/pai/pull/409#discussion_r176959993) [](ancestors = 176959993)

File: job-tutorial/README.md [hunk 5]
```diff
@@ -104,6 +111,10 @@ Below please find the detailed explanation for each of the parameters in the con
 | `taskRole.cpuNumber`           | Integer, required          | CPU number for one task in the task role, no less than 1 |
 | `taskRole.memoryMB`            | Integer, required          | Memory for one task in the task role, no less than 100 |
 | `taskRole.gpuNumber`           | Integer, required          | GPU number for one task in the task role, no less than 0 |
+| `taskRole.portList`            | List, optional             | List of `portType` to use                |
+| `taskRole.portType.label`      | String in `^[A-Za-z0-9\-._~]+$` format, required | Label name for the port type |
+| `taskRole.portType.startAt`    | Integer, required          | The port no to start with in the port type, 0 for random selection |
```
Comment (qinchen123) [inline job-tutorial/README.md]: >port no [](start = 68, length = 7)

port no ? port

**Pull request merged**